rust/src/table/convert_to_delta.rs

```diff
@@ -0,0 +1,277 @@
+            if let Ok(meta) = path {
+                if Some("parquet") == meta.location.extension() {
+                    debug!("Found parquet file in {:#?}", meta.location);
+                    Some(meta)
+                } else {
+                    None
+                }
+            } else {
+                None
+            }
+        })
+        .collect()
+        .await)
+}
+
+async fn arrow_schema(
+    file: &ObjectMeta,
+    store: Arc<DeltaObjectStore>,
+) -> Result<ArrowSchema, Error> {
+    Ok(
+        ParquetRecordBatchStreamBuilder::new(ParquetObjectReader::new(store.clone(), file.clone()))
+            .await?
+            .schema()
+            .as_ref()
+            .clone(),
+    )
+}
+
+fn add_action(
+    file: &ObjectMeta,
+    partition_values: HashMap<String, Option<String>>,
+) -> Result<Action, Error> {
+    Ok(Action::add(Add {
+        path: file.location.to_string(),
+        size: i64::try_from(file.size)?,
+        partition_values,
+        modification_time: file.last_modified.timestamp_millis(),
+        ..Default::default()
+    }))
+}
+
+fn partitions(
+    file: &ObjectMeta,
+    // A HashSet for all unique partition columns in a Parquet table
+    partition_columns: &mut HashSet<String>,
+    // A Vector for schema fields of all unique partition columns in a Parquet table
+    partition_schema_fields: &mut Vec<SchemaField>,
+) -> Result<HashMap<String, Option<String>>, Error> {
+    // A HashMap from partition column to value for this parquet file
+    let mut partitions = HashMap::new();
+    for s in file.location.as_ref().split('/') {
+        if !s.ends_with(".parquet") {
+            let partition = DeltaTablePartition::try_from(s)?;
+            let (key, val) = (partition.key.to_string(), partition.value.to_string());
+            debug!(
+                "Found partition {partition:#?} in parquet file {:#?}",
+                file.location
+            );
+            partitions.insert(key.clone(), Some(val));
+            if partition_columns.insert(key.clone()) {
+                partition_schema_fields.push(SchemaField::new(
+                    key,
+                    SchemaDataType::primitive("string".to_string()),
+                    true,
+                    HashMap::new(),
+                ));
+            }
+        }
+    }
+    Ok(partitions)
+}
+
+async fn delta_table(
+    files: Vec<ObjectMeta>,
+    store: Arc<DeltaObjectStore>,
+    url: Url,
+) -> Result<DeltaTable, Error> {
+    let (mut schemas, mut actions, mut partition_columns, mut partition_schema_fields) =
+        (Vec::new(), Vec::new(), HashSet::new(), Vec::new());
+    for file in files {
+        debug!("Processing parquet file: {:#?}", file.location);
+        schemas.push(arrow_schema(&file, store.clone()).await?);
+        let partition_values =
+            partitions(&file, &mut partition_columns, &mut partition_schema_fields)?;
+        actions.push(add_action(&file, partition_values)?);
+    }
```

Comment:
Based on my read of the module added, this code will read every single `.parquet` file that exists int he configured object store and mash their schemas together. Is there particular concern taking the latest file and calling that the table's schema?

Reply:
Yes, this code will read all parquet files and smash their schemas together. This is because timestamp will not be preserved when copying files in S3. A newer timestamp does not necessarily mean a newer version of the parquet file.
Some discussions https://github.com/aws/aws-cli/issues/3415, https://github.com/aws/aws-cli/discussions/7071.